// react-native-atoz-listview/components/SectionHeader.js
'use strict';

import PropTypes from 'prop-types';
import React, {
  Component,
} from 'react';
import ReactNative, {
  StyleSheet,
  View,
  Text
} from 'react-native';

export default class SectionHeader extends Component {

  componentDidMount() {
    this.props.updateTag && this.props.updateTag(ReactNative.findNodeHandle(this.refs.view), this.props.sectionId);
  }

  render() {
    const SectionComponent = this.props.component;
    const content = SectionComponent ?
      <SectionComponent {...this.props} /> :
      <Text style={{fontFamily:'Montserrat-Regular',fontSize:22, marginLeft:20, marginTop:20, marginBottom:10}}>{this.props.title}</Text>;
    return (
      <View ref="view">
        {content}
      </View>
    );
  }
}

const styles = StyleSheet.create({
  container: {
    backgroundColor: '#f8f8f8',
    borderTopWidth: 1,
    borderTopColor: '#ececec',
  },
  text: {
    fontWeight: '700',
    paddingTop: 2,
    paddingBottom: 2,
    paddingLeft: 2,
  }
});

SectionHeader.propTypes = {

  /**
   * The id of the section
   */
  sectionId: PropTypes.oneOfType([
    PropTypes.number,
    PropTypes.string
  ]),

  /**
   * A component to render for each section item
   */
  component: PropTypes.func,

  /**
   * A function used to propagate the root nodes handle back to the parent
   */
  updateTag: PropTypes.func
};
